File: app/src/main/java/com/tv/meplayer/data/StorageService.kt
package com.tv.meplayer.data

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

/**
 * 存储服务，对应 Flutter 的 `storage_service.dart`。
 *
 * 基于 SharedPreferences + JSON 序列化，提供跨 App 全局持久化的 key-value 存储。
 */
object StorageService {

    private lateinit var prefs: android.content.SharedPreferences

    fun init(context: Context, boxName: String = "flutter_player_android") {
        prefs = context.getSharedPreferences(boxName, Context.MODE_PRIVATE)
    }

    // ── 基础读写 ─────────────────────────────────────────────────────────

    fun getString(key: String, default: String = ""): String =
        prefs.getString(key, default) ?: default

    fun setString(key: String, value: String) {
        prefs.edit().putString(key, value).apply()
    }

    fun getInt(key: String, default: Int = 0): Int =
        prefs.getInt(key, default)

    fun setInt(key: String, value: Int) {
        prefs.edit().putInt(key, value).apply()
    }

    fun getBoolean(key: String, default: Boolean = false): Boolean =
        prefs.getBoolean(key, default)

    fun setBoolean(key: String, value: Boolean) {
        prefs.edit().putBoolean(key, value).apply()
    }

    fun getLong(key: String, default: Long = 0L): Long =
        prefs.getLong(key, default)

    fun setLong(key: String, value: Long) {
        prefs.edit().putLong(key, value).apply()
    }

    // ── List<String> ────────────────────────────────────────────────────

    fun getStringList(key: String, default: List<String> = emptyList()): List<String> {
        val raw = prefs.getString(key, null) ?: return default
        return runCatching {
            JSONArray(raw).let { arr ->
                (0 until arr.length()).map { arr.getString(it) }
            }
        }.getOrDefault(default)
    }

    fun setStringList(key: String, value: List<String>) {
        val arr = JSONArray()
        value.forEach { arr.put(it) }
        prefs.edit().putString(key, arr.toString()).apply()
    }

    // ── List<Map<String,String>> ────────────────────────────────────────

    fun getMapList(key: String, default: List<Map<String, String>> = emptyList()): List<Map<String, String>> {
        val raw = prefs.getString(key, null) ?: return default
        return runCatching {
            JSONArray(raw).let { arr ->
                (0 until arr.length()).map { idx ->
                    val obj = arr.getJSONObject(idx)
                    val map = mutableMapOf<String, String>()
                    obj.keys().forEach { k -> map[k] = obj.getString(k) }
                    map
                }
            }
        }.getOrDefault(default)
    }

    fun setMapList(key: String, value: List<Map<String, String>>) {
        val arr = JSONArray()
        value.forEach { m ->
            val obj = JSONObject()
            m.forEach { (k, v) -> obj.put(k, v) }
            arr.put(obj)
        }
        prefs.edit().putString(key, arr.toString()).apply()
    }

    // ── JSON 对象（用于扫描结果缓存） ────────────────────────────────────

    fun getJsonArray(key: String, default: String = "[]"): String =
        prefs.getString(key, default) ?: default

    fun setJsonArray(key: String, json: String) {
        prefs.edit().putString(key, json).apply()
    }

    fun delete(key: String) {
        prefs.edit().remove(key).apply()
    }

    // ── Key 常量（与 Flutter 端完全一致） ───────────────────────────────

    const val kHardwareDecode = "hw_decode"
    const val kMpvProfile = "mpv_profile"
    const val kCompatMode = "compat_mode"
    const val kIptvSources = "iptv_sources"
    const val kVideoScanPaths = "video_scan_paths"
    const val kMusicScanPaths = "music_scan_paths"
    const val kRecentFiles = "recent_files"
    const val kPlayerVolume = "player_volume"
    const val kPlayerBackend = "player_backend"

    const val kThemeMode = "theme_mode"
    const val kSeedColor = "seed_color"

    const val kVideoSortMode = "video_sort_mode"
    const val kVideoThumbCache = "video_thumb_cache"

    const val kMusicCategory = "music_category"
    const val kMusicSongSort = "music_song_sort"
    const val kMusicGroupSort = "music_group_sort"
    const val kMusicLibraryCache = "music_library_cache"

    const val kVideoCategory = "video_category"
    const val kVideoSortAndroid = "video_sort_android"
    const val kVideoFolderSort = "video_folder_sort"
    const val kVideoLibraryCache = "video_library_cache_android"
    const val kMusicLibraryCacheAndroid = "music_library_cache_android"
}

File: Services/StorageService.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using Microsoft.Data.Sqlite;

namespace MEPlayer.Services;

/// <summary>
/// 基于 SQLite 的键值存储，替代 Flutter 端的 Hive。
///
/// 实现说明：
/// - 所有键值对存放在 SQLite 数据库的 kv 表（key TEXT PRIMARY KEY, value TEXT）。
/// - value 字段保存 value 的 JSON 序列化字符串（基本类型也以 JSON 文本存储，
///   反序列化时按目标 T 类型还原）。
/// - 单一 SqliteConnection 贯穿 App 生命周期，所有操作加锁串行化，
///   并启用 WAL 模式以兼顾读性能与崩溃恢复。
/// - 保留与原 JSON 版本完全相同的公开接口（Init / Get / Set / Delete），
///   调用方无需任何改动。
/// </summary>
public sealed class StorageService : IDisposable
{
    private string _dbPath = "";
    private readonly object _lock = new();
    private SqliteConnection? _conn;
    private bool _disposed;

    private static readonly JsonSerializerOptions _opts = new()
    {
        WriteIndented = false,
    };

    public void Init(string path)
    {
        // 兼容旧调用：原 path 形如 .../settings.json，这里改为同目录下 settings.db。
        // 如果传入的已经是 .db 路径则直接使用。
        _dbPath = Path.GetExtension(path).Equals(".json", StringComparison.OrdinalIgnoreCase)
            ? Path.ChangeExtension(path, ".db")
            : path;

        lock (_lock)
        {
            try
            {
                var dir = Path.GetDirectoryName(_dbPath);
                if (!string.IsNullOrEmpty(dir) && !Directory.Exists(dir))
                    Directory.CreateDirectory(dir);

                _conn = new SqliteConnection($"Data Source={_dbPath}");
                _conn.Open();

                using (var cmd = _conn.CreateCommand())
                {
                    cmd.CommandText =
                        "PRAGMA journal_mode=WAL;" +
                        "PRAGMA synchronous=NORMAL;" +
                        "CREATE TABLE IF NOT EXISTS kv (" +
                        "  key TEXT PRIMARY KEY NOT NULL," +
                        "  value TEXT NOT NULL" +
                        ");" +
                        "CREATE INDEX IF NOT EXISTS idx_kv_key ON kv(key);";
                    cmd.ExecuteNonQuery();
                }
            }
            catch
            {
                // 初始化失败时保留 _conn=null，后续 Get/Set 会走默认值/无操作兜底，
                // 避免存储故障导致整个 App 无法启动。
                _conn = null;
            }
        }
    }

    public T Get<T>(string key, T defaultValue)
    {
        lock (_lock)
        {
            if (_conn == null) return defaultValue;
            try
            {
                using var cmd = _conn.CreateCommand();
                cmd.CommandText = "SELECT value FROM kv WHERE key = @k";
                cmd.Parameters.AddWithValue("@k", key);
                var raw = cmd.ExecuteScalar() as string;
                if (string.IsNullOrEmpty(raw)) return defaultValue;
                return JsonSerializer.Deserialize<T>(raw, _opts) ?? defaultValue;
            }
            catch
            {
                return defaultValue;
            }
        }
    }

    public void Set<T>(string key, T value)
    {
        lock (_lock)
        {
            if (_conn == null) return;
            try
            {
                var json = JsonSerializer.Serialize(value, value?.GetType() ?? typeof(T), _opts);
                using var cmd = _conn.CreateCommand();
                cmd.CommandText =
                    "INSERT INTO kv(key, value) VALUES(@k, @v) " +
                    "ON CONFLICT(key) DO UPDATE SET value = excluded.value";
                cmd.Parameters.AddWithValue("@k", key);
                cmd.Parameters.AddWithValue("@v", json);
                cmd.ExecuteNonQuery();
            }
            catch
            {
                // 写入失败不影响运行
            }
        }
    }

    public void Delete(string key)
    {
        lock (_lock)
        {
            if (_conn == null) return;
            try
            {
                using var cmd = _conn.CreateCommand();
                cmd.CommandText = "DELETE FROM kv WHERE key = @k";
                cmd.Parameters.AddWithValue("@k", key);
                cmd.ExecuteNonQuery();
            }
            catch
            {
                // 忽略
            }
        }
    }

    public bool HasKey(string key)
    {
        lock (_lock)
        {
            if (_conn == null) return false;
            try
            {
                using var cmd = _conn.CreateCommand();
                cmd.CommandText = "SELECT 1 FROM kv WHERE key = @k LIMIT 1";
                cmd.Parameters.AddWithValue("@k", key);
                return cmd.ExecuteScalar() != null;
            }
            catch
            {
                return false;
            }
        }
    }

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        lock (_lock)
        {
            try { _conn?.Dispose(); } catch { }
            _conn = null;
        }
    }

    // ── 共享 key 常量（与 Flutter 端一一对应） ──────────────────────────────
    public const string KHardwareDecode = "hw_decode";
    public const string KMpvProfile = "mpv_profile";
    public const string KCompatMode = "compat_mode";
    public const string KIptvSources = "iptv_sources";
    public const string KVideoScanPaths = "video_scan_paths";
    public const string KMusicScanPaths = "music_scan_paths";
    public const string KRecentFiles = "recent_files";
    public const string KPlayerVolume = "player_volume";

    public const string KThemeMode = "theme_mode";
    public const string KSeedColor = "seed_color";

    public const string KVideoSortMode = "video_sort_mode";
    public const string KVideoThumbCache = "video_thumb_cache";

    public const string KMusicCategory = "music_category";
    public const string KMusicSongSort = "music_song_sort";
    public const string KMusicGroupSort = "music_group_sort";
    public const string KMusicLibraryCache = "music_library_cache";
}
